usr.bin/talk/talk.1
.\"	$OpenBSD: talk.1,v 1.27 2017/05/25 20:25:50 tedu Exp $
.\"	$NetBSD: talk.1,v 1.3 1994/12/09 02:14:23 jtc Exp $
.\"
.\" Copyright (c) 1983, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)talk.1	8.1 (Berkeley) 6/6/93
.\"
.Dd $Mdocdate: May 25 2017 $
.Dt TALK 1
.Os
.Sh NAME
.Nm talk
.Nd talk to another user
.Sh SYNOPSIS
.Nm talk
.Op Fl Hs
.Ar person
.Op Ar ttyname
.Sh DESCRIPTION
.Nm
is a visual communication program which copies lines from your
terminal to that of another user.
.Pp
The command arguments are as follows:
.Bl -tag -width ttyname
.It Fl H
Don't escape characters with the high bit set.
This may be useful for certain character sets, but could cause erratic
behaviour on some terminals.
.It Fl s
Use smooth scrolling in the
.Nm
window.
The default is to clear the next two rows and jump from the bottom of
the window to the top.
.It Ar person
If you wish to talk to someone on your own machine, then
.Ar person
is just the person's login name.
If you wish to talk to a user on another host, then
.Ar person
is of the form
.Ql user@host .
.It Ar ttyname
If you wish to talk to a user who is logged in more than once, the
.Ar ttyname
argument may be used to indicate the appropriate terminal
name, where
.Ar ttyname
is of the form
.Ql ttyXX .
.El
.Pp
When first called,
.Nm
sends the message
.Bd -literal -offset indent
Message from Talk_Daemon@localhost...
talk: connection requested by your_name@your_machine.
talk: respond with: talk your_name@your_machine
.Ed
.Pp
to the user you wish to talk to.
At this point, the recipient of the message should reply by typing
.Pp
.Dl $ talk \ your_name@your_machine
.Pp
It doesn't matter from which machine the recipient replies, as
long as the login name is the same.
If the machine is not the one to which
the talk request was sent, it is noted on the screen.
Once communication is established,
the two parties may type simultaneously, with their output appearing
in separate windows.
Typing control-L
.Pq Ql ^L
will cause the screen to
be reprinted, while the erase, kill, and word kill characters will
behave normally.
To exit, just type the interrupt character;
.Nm
then moves the cursor to the bottom of the screen and restores the
terminal to its previous state.
.Pp
Permission to talk may be denied or granted by use of the
.Xr mesg 1
command.
At the outset talking is allowed.
Certain commands, such as
.Xr pr 1 ,
disallow messages in order to
prevent messy output.
.Sh ASYNCHRONOUS EVENTS
.Bl -tag -width SIGINTXXX
.It Dv SIGINT
Terminate
.Nm
and exit with a zero status.
.El
.Sh FILES
.Bl -tag -width /var/run/utmp -compact
.It Pa /etc/hosts
to find the recipient's machine
.It Pa /var/run/utmp
to find the recipient's tty
.El
.Sh EXIT STATUS
The
.Nm
utility exits 0 on success, and >0 if either an error occurred or
.Nm
is
invoked on an unsupported terminal.
.Sh SEE ALSO
.Xr mail 1 ,
.Xr mesg 1 ,
.Xr who 1 ,
.Xr write 1 ,
.Xr talkd 8
.Sh STANDARDS
The
.Nm
utility is compliant with the
.St -p1003.1-2008
specification,
though its presence is optional.
.Pp
The flags
.Op Fl Hs
are extensions to that specification.
.Sh HISTORY
The
.Nm
command appeared in
.Bx 4.2 .


usr.bin/talk/talk.c
/*	$OpenBSD: talk.c,v 1.12 2016/03/08 20:07:46 mestre Exp $	*/
/*	$NetBSD: talk.c,v 1.3 1994/12/09 02:14:25 jtc Exp $	*/

/*
 * Copyright (c) 1983, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <err.h>
#include <string.h>
#include <unistd.h>

#include "talk.h"
#include "talk_ctl.h"

/*
 * talk:	A visual form of write. Using sockets, a two way
 *		connection is set up between the two people talking.
 *		With the aid of curses, the screen is split into two
 *		windows, and each users text is added to the window,
 *		one character at a time...
 *
 *		Written by Kipp Hickman
 *		
 *		Modified to run under 4.1a by Clem Cole and Peter Moore
 *		Modified to run between hosts by Peter Moore, 8/19/82
 *		Modified to run under 4.1c by Peter Moore 3/17/83
 */

int
main(int argc, char *argv[])
{
	if (pledge("stdio rpath inet dns getpw tty", NULL) == -1)
		err(1, "pledge");

	get_names(argc, argv);
	init_display();
	open_ctl();
	open_sockt();
	start_msgs();
	if (!check_local())
		invite_remote();
	end_msgs();
	set_edit_chars();

	if (his_machine_addr.s_addr == my_machine_addr.s_addr) {
		if (pledge("stdio tty", NULL) == -1)
			err(1, "pledge");
	} else {
		if (pledge("stdio inet tty", NULL) == -1)
			err(1, "pledge");
	}

	talk();
	return (0);
}
